relentless/optimize/method.py

```diff
@@ -174,27 +186,145 @@ def optimize(self, objective):
+        self._max_iter = value
+
     @property
     def step_size(self):
-        """float: The step size hyperparameter (:math:`\alpha`)."""
+        r"""float or :class:`LineSearch`: The step size hyperparameter (:math:`\alpha`)."""
         return self._step_size
 
     @step_size.setter
     def step_size(self, value):
-        if value <= 0:
-            raise ValueError('The step size must be positive.')
+        try:
+            if value <= 0:
+                raise ValueError('The step size must be positive if defined as a float.')
+        except TypeError:
+            if not isinstance(value, LineSearch):
+                raise ValueError('The step size must be a float or an instance of LineSearch.')
         self._step_size = value
 
+    @property
+    def line_search(self):
+        """:class:`LineSearch`: The line search object used to optimize the step size."""
+        return self._line_search
+
+    @line_search.setter
+    def line_search(self, value):
+        if value is not None and not isinstance(value, LineSearch):
+            raise TypeError('If defined, the line search parameter must be a LineSearch object.')
+        self._line_search = value
+
+#TODO: document
+class LineSearch:
+    """
+    """
+    def __init__(self, abs_tol, max_iter):
+        self.abs_tol = abs_tol
+        self.max_iter = max_iter
+
+    def find(self, objective, res_start, res_end):
+        """
+        """
+        res_0 = objective.compute()
+        ovars = res_0.design_variables
+
+        # compute normalized search direction, adjust variables based on max step size
+        d = {x: res_end.design_variables[x]-res_start.design_variables[x] for x in ovars}
+        max_step = np.linalg.norm(np.asarray(list(d.values())))
+        for x in ovars:
+            d[x] /= max_step
+            x.value += max_step*d[x]
+        res = objective.compute()
+
+        # compute initial and post-adjustment target values
+        target_0 = 0
+        target = 0
+        for x in ovars:
+            target_0 += (d[x]*-res_0.gradient(x))
+            target += (d[x]*-res.gradient(x))
+
+        # check if max step size is acceptable
+        if target > 0 or np.abs(target) < self.abs_tol: #combine conditions as (target > -self.abs_tol) ?
+            for x in ovars:
+                x.value = ovars[x]
+            return res
+
+        # iterate to minimize target
+        else:
+            rstep = np.array([0, max_step])
+            rtarget = np.array([target_0, target])
+            iter_num = 0
+            while np.abs(target) > self.abs_tol and iter_num < self.max_iter:
+                # linear interpolation for step size
+                step = (rstep[0]*rtarget[1] - rstep[1]*rtarget[0])/(rtarget[1] - rtarget[0])
+
+                # adjust variables based on new step size, compute target
+                for x in ovars:
+                    x.value = ovars[x] + step*d[x]
+                res = objective.compute()
+                target = 0
```

Comment:
Call this like `new_target` or `cur_target` or something to distinguish it from all the other ones above.